pex/pip/log_analyzer.py

```diff
@@ -92,21 +92,22 @@ def _check_returncode(self, stderr=None):
         ]
         if activated_analyzers:
             collected = []
-            with open(self._log, "r") as fp:
-                for line in fp:
-                    if not activated_analyzers:
-                        break
-                    for index, analyzer in enumerate(activated_analyzers):
-                        result = analyzer.analyze(line)
-                        if isinstance(result.data, ErrorMessage):
-                            collected.append(result.data)
-                        if isinstance(result, LogAnalyzer.Complete):
-                            activated_analyzers.pop(index).analysis_completed()
-                            if not activated_analyzers:
-                                break
-            for analyzer in activated_analyzers:
-                analyzer.analysis_completed()
-            if not self._preserve_log:
-                os.unlink(self._log)
-            stderr = (stderr or b"") + "".join(collected).encode("utf-8")
+            if os.path.isfile(self._log):
```

Comment:
I'm claiming this is a bug. If there are activated analyzers there must be a log for them to analyze. If there is not, that's an error - or should be. That error is now masked by silently doing a shoulder shrug afaict.

Reply:
If there is some superclass contract wonkiness (or dressed-up "subtletly"), maybe that can be called out here. "We do this crazy thing because of this other crazy thing ...".

Reply:
Not sure I follow. If the interpreter is busted in some way (e.g., the distutils issue) then it will error out before pip has a chance to create the log. That seems to me to be the real error, which was previously masked by the log-reading error, but is now allowed to express itself. 

I'm not sure what the wonkiness is here? I read this code as "when the process exits, analyze its log for useful info, and then propagate its exit code, augmented with this useful info". If there is no log, there is nothing to augment it with, but we still propagate the exit code.  I guess the wonky case might be a process that succeeds but write no log? We can check for that.

Reply:
The wonkiness is that in the small there is nothing wrong. We have analyzers, they want to analyze, there should be a log for them to analyze. If there is not a log and something went wrong somewhere else, its very odd that control gets back here with no signal saying - things failed. You're taking the log file not existing as that signal. Perhaps that fine if that's always true. Maybe at least call it out in a comment? It feels like there is more than one layer involved here and the current layer is having to know something about other layers to make the bold claim that the file that is its reason for existence is not available but that's ok.

Reply:
I've read more through the surrounding code an a comment is likely the best that can be done. I'm happy following up and adding that.

Reply:
I still don't really understand your point. "there should be a log for them to analyze" was a previous assumption, but it turns out to be incorrect. Sometimes there will be no log, if the pip process fails early enough. The failure of that process is the one that is useful to show to the user, not the absence of a log file. 

To put it another way, what if instead of no log there were an empty log? This PR wouldn't have been necessary, the analyzers would operate vacuously, but the exact same thing would happen. 

I'll add a comment, but feel free to follow up with a better one.